File: benchmarks/tsb/bench_where_mask_series_fn.ts
/**
 * Benchmark: whereSeries / maskSeries — standalone functional where/mask for Series.
 * Outputs JSON: {"function": "where_mask_series_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, whereSeries, maskSeries } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

const s = new Series({ data: Array.from({ length: SIZE }, (_, i) => i * 0.1) });
const cond = (v: unknown) => (v as number) > SIZE * 0.05;
const condArr = Array.from({ length: SIZE }, (_, i) => i > SIZE * 0.5);

for (let i = 0; i < WARMUP; i++) {
  whereSeries(s, cond, 0);
  maskSeries(s, condArr, -1);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  whereSeries(s, cond, 0);
  maskSeries(s, condArr, -1);
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "where_mask_series_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_where_mask_series_fn.py
"""
Benchmark: pandas Series.where() / Series.mask() — conditional replacement.
Outputs JSON: {"function": "where_mask_series_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

s = pd.Series([i * 0.1 for i in range(SIZE)])
cond = s > SIZE * 0.05
cond_arr = pd.Series([i > SIZE * 0.5 for i in range(SIZE)])

for _ in range(WARMUP):
    s.where(cond, 0)
    s.mask(cond_arr, -1)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.where(cond, 0)
    s.mask(cond_arr, -1)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "where_mask_series_fn", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))
